File: src/backup/gateway/collectors/macro_collector.py
"""Macro collection adapter.

Calls all 8 macro collectors via asyncio.gather() and aggregates
their outputs into a single MacroResult.
"""

from __future__ import annotations

import asyncio
import time
from datetime import UTC, datetime
from typing import Any, Optional

from engine.macro.collectors.calendar.collector import CalendarCollector
from engine.macro.collectors.central_bank.collector import CentralBankCollector
from engine.macro.collectors.cot.collector import COTCollector
from engine.macro.collectors.dxy.collector import DXYCollector
from engine.macro.collectors.economic_data.collector import EconomicDataCollector
from engine.macro.collectors.intermarket.collector import IntermarketCollector
from engine.macro.collectors.news.collector import NewsCollector
from engine.macro.collectors.sentiment.collector import SentimentCollector
from engine.shared.logging import get_logger
from gateway.constants import PipelineStage
from gateway.context.models import MacroResult
from gateway.observability.metrics import (
    GATEWAY_MACRO_COLLECT_DURATION,
    GATEWAY_STAGE_ERRORS,
)

logger = get_logger(__name__)


class MacroCollector:
    """Collects all macro data in parallel and returns a unified result."""

    def __init__(
        self,
        *,
        cb_collector: CentralBankCollector,
        cot_collector: COTCollector,
        economic_collector: EconomicDataCollector,
        news_collector: NewsCollector,
        calendar_collector: CalendarCollector,
        dxy_collector: DXYCollector,
        intermarket_collector: IntermarketCollector,
        sentiment_collector: SentimentCollector,
    ) -> None:
        self._collectors: dict[str, Any] = {
            "central_bank": cb_collector,
            "cot": cot_collector,
            "economic": economic_collector,
            "news": news_collector,
            "calendar": calendar_collector,
            "dxy": dxy_collector,
            "intermarket": intermarket_collector,
            "sentiment": sentiment_collector,
        }

    async def collect(
        self,
        *,
        trace_id: Optional[str] = None,
    ) -> MacroResult:
        """Run all 8 macro collectors in parallel."""
        start = time.monotonic()
        names = list(self._collectors.keys())
        collectors = list(self._collectors.values())

        tasks = [c.collect() for c in collectors]
        raw_results = await asyncio.gather(*tasks, return_exceptions=True)

        datasets: dict[str, Optional[dict]] = {}
        errors: dict[str, str] = {}

        for name, result in zip(names, raw_results):
            if isinstance(result, Exception):
                logger.error(
                    "macro_collector_failed",
                    extra={
                        "collector": name,
                        "error": str(result),
                        "trace_id": trace_id,
                    },
                )
                GATEWAY_STAGE_ERRORS.labels(
                    stage=PipelineStage.MACRO_COLLECTOR,
                    error_type=type(result).__name__,
                ).inc()
                datasets[name] = None
                errors[name] = str(result)
            else:
                datasets[name] = self._serialize(result)

        elapsed_ms = (time.monotonic() - start) * 1000
        GATEWAY_MACRO_COLLECT_DURATION.observe(elapsed_ms / 1000)

        logger.info(
            "macro_collection_completed",
            extra={
                "datasets_available": [
                    n for n, d in datasets.items() if d is not None
                ],
                "datasets_failed": list(errors.keys()),
                "duration_ms": round(elapsed_ms, 1),
                "trace_id": trace_id,
            },
        )

        return MacroResult(
            central_bank=datasets.get("central_bank"),
            cot=datasets.get("cot"),
            economic=datasets.get("economic"),
            news=datasets.get("news"),
            calendar=datasets.get("calendar"),
            dxy=datasets.get("dxy"),
            intermarket=datasets.get("intermarket"),
            sentiment=datasets.get("sentiment"),
            collected_at=datetime.now(UTC),
            duration_ms=elapsed_ms,
            errors=errors,
        )

    @staticmethod
    def _serialize(result: Any) -> dict:
        """Safely serialize a collector result to a JSON-compatible dict."""
        if isinstance(result, dict):
            return result
        if hasattr(result, "model_dump"):
            return result.model_dump(mode="json")
        return {"raw": str(result)}

File: src/gateway/internal/collectors/macro_collector.go
package collectors

import (
	"context"
	"encoding/json"
	"time"

	"github.com/rs/zerolog"

	"github.com/flamegreat-1/etradie/src/gateway/internal/constants"
	"github.com/flamegreat-1/etradie/src/gateway/internal/infra"
	"github.com/flamegreat-1/etradie/src/gateway/internal/models"
	"github.com/flamegreat-1/etradie/src/gateway/internal/observability"
)

// MacroCollector calls the Python engine to collect all 8 macro datasets
// via a single HTTP call (the Python side runs them in parallel).
// Results are cached in Redis to avoid redundant HTTP calls within
// the configured TTL window.
type MacroCollector struct {
	engine   *infra.EngineHTTPClient
	redis    *infra.RedisClient
	cacheTTL time.Duration
	log      zerolog.Logger
}

// NewMacroCollector creates a MacroCollector backed by the engine HTTP client
// with optional Redis caching. If redis is nil or TTL is 0, caching is disabled.
func NewMacroCollector(engine *infra.EngineHTTPClient, redis *infra.RedisClient, cacheTTLSeconds int) *MacroCollector {
	return &MacroCollector{
		engine:   engine,
		redis:    redis,
		cacheTTL: time.Duration(cacheTTLSeconds) * time.Second,
		log:      observability.Logger("macro_collector"),
	}
}

// Collect runs all 8 macro collectors via the Python engine and returns a MacroResult.
func (c *MacroCollector) Collect(ctx context.Context, traceID string) (*models.MacroResult, error) {
	// Check global cache first.
	if cached := c.getFromCache(ctx, traceID); cached != nil {
		return cached, nil
	}

	start := time.Now()

	reqBody := map[string]interface{}{
		"trace_id": traceID,
	}

	resp, err := c.engine.PostJSON(ctx, "/internal/macro/collect", reqBody)

	elapsedMs := float64(time.Since(start).Milliseconds())
	observability.GatewayMacroCollectDuration.Observe(time.Since(start).Seconds())

	if err != nil {
		observability.GatewayStageErrors.WithLabelValues(
			constants.StageMacroCollector.String(), "http_error",
		).Inc()
		c.log.Error().
			Err(err).
			Str("trace_id", traceID).
			Msg("macro_collection_http_failed")
		return nil, err
	}

	result := &models.MacroResult{
		CentralBank: getDatasetMap(resp, "central_bank"),
		COT:         getDatasetMap(resp, "cot"),
		Economic:    getDatasetMap(resp, "economic"),
		News:        getDatasetMap(resp, "news"),
		Calendar:    getDatasetMap(resp, "calendar"),
		DXY:         getDatasetMap(resp, "dxy"),
		Intermarket: getDatasetMap(resp, "intermarket"),
		Sentiment:   getDatasetMap(resp, "sentiment"),
		CollectedAt: time.Now().UTC(),
		DurationMs:  elapsedMs,
		Errors:      getErrorsMap(resp),
	}

	available := result.AvailableDatasets()
	failed := make([]string, 0, len(result.Errors))
	for k := range result.Errors {
		failed = append(failed, k)
	}

	c.log.Info().
		Strs("datasets_available", available).
		Strs("datasets_failed", failed).
		Float64("duration_ms", elapsedMs).
		Str("trace_id", traceID).
		Msg("macro_collection_completed")

	// Cache successful result globally.
	c.storeInCache(ctx, result, traceID)

	return result, nil
}

// macroCacheKey returns the global cache key for macro results.
func macroCacheKey() string {
	return "latest"
}

func (c *MacroCollector) getFromCache(ctx context.Context, traceID string) *models.MacroResult {
	if c.redis == nil || c.cacheTTL <= 0 {
		return nil
	}

	raw, err := c.redis.GetRaw(ctx, constants.GatewayCacheNamespace, constants.MacroResultCacheKeyPrefix+":"+macroCacheKey())
	if err != nil {
		c.log.Warn().Err(err).Str("trace_id", traceID).Msg("macro_cache_read_error")
		return nil
	}
	if raw == nil {
		return nil
	}

	// Single-pass unmarshal directly into the typed struct.
	var result models.MacroResult
	if err := json.Unmarshal(raw, &result); err != nil {
		c.log.Warn().Err(err).Str("trace_id", traceID).Msg("macro_cache_unmarshal_error")
		return nil
	}

	c.log.Info().
		Str("trace_id", traceID).
		Msg("macro_result_served_from_cache")
	return &result
}

func (c *MacroCollector) storeInCache(ctx context.Context, result *models.MacroResult, traceID string) {
	if c.redis == nil || c.cacheTTL <= 0 {
		return
	}

	if err := c.redis.Set(ctx, constants.GatewayCacheNamespace, constants.MacroResultCacheKeyPrefix+":"+macroCacheKey(), result, c.cacheTTL); err != nil {
		c.log.Warn().Err(err).Str("trace_id", traceID).Msg("macro_cache_write_error")
	}
}

func getDatasetMap(resp map[string]interface{}, key string) map[string]interface{} {
	v, ok := resp[key]
	if !ok || v == nil {
		return nil
	}
	m, ok := v.(map[string]interface{})
	if !ok {
		return nil
	}
	return m
}

func getErrorsMap(resp map[string]interface{}) map[string]string {
	v, ok := resp["errors"]
	if !ok || v == nil {
		return make(map[string]string)
	}
	raw, ok := v.(map[string]interface{})
	if !ok {
		return make(map[string]string)
	}
	out := make(map[string]string, len(raw))
	for k, val := range raw {
		if s, ok := val.(string); ok {
			out[k] = s
		}
	}
	return out
}
